File: tests/attack/whitebox/test_fgsm.py
import torch

from pyadv import config_parser, get_attacker, get_dataset, get_model, logger


def test_fgsm():
    setting = {
        "attacker": "FGSM",
        "norm": "Linf",
        "epsilon": 0.015625,
        "criterion": "cw",
        "dataset": "cifar10",
        "n_examples": 100,
        "model": "Carmon2019Unlabeled",
        "batch_size": 100,
    }
    config_parser().clear()
    config = config_parser(setting)
    if torch.backends.cudnn.is_available():
        config.device = "cuda"
    elif torch.backends.mps.is_available():
        config.device = "mps"
    else:
        raise RuntimeError("No GPU found")

    logger.setLevel("DEBUG")
    attacker = get_attacker()
    model, transform = get_model()
    data, label = get_dataset(transform)

    attacker.attack(model, data, label)

File: tests/test_fgsm.cpp
#include <advlib.hpp>
#include <iostream>

int main() {
    try {
        auto setting = advlib::load_yaml("tests/example/config.yaml");
        auto config = advlib::Config(setting);
        std::cout << config.value() << std::endl;
        auto model = advlib::get_model(config);
        auto [data, label] = advlib::get_dataset(config);
        return 0;
    } catch (const std::exception &e) {
        std::cerr << e.what() << std::endl;
        return 1;
    }
}
